proto/proto.go

```diff
@@ -0,0 +1,256 @@
+package proto
+
+import (
+	"fmt"
+	"reflect"
+	"sync/atomic"
+	"unsafe"
+)
+
+func Size(v interface{}) int {
+	t, p := inspect(v)
+	c := cachedCodecOf(t)
+	return c.size(p, inline)
+}
+
+func Marshal(v interface{}) ([]byte, error) {
+	t, p := inspect(v)
+	c := cachedCodecOf(t)
+	b := make([]byte, c.size(p, inline))
+	_, err := c.encode(b, p, inline)
+	if err != nil {
+		return nil, fmt.Errorf("proto.Marshal(%T): %w", v, err)
+	}
+	return b, nil
+}
+
+func MarshalTo(b []byte, v interface{}) (int, error) {
+	t, p := inspect(v)
+	c := cachedCodecOf(t)
+	n, err := c.encode(b, p, inline)
+	if err != nil {
+		err = fmt.Errorf("proto.MarshalTo: %w", err)
+	}
+	return n, err
+}
+
+func Unmarshal(b []byte, v interface{}) error {
+	if len(b) == 0 {
+		return nil
```

Comment:
Shouldn't this be an error? or at least this check omitted, so that the codec has the opportunity to handle the empty case?

Reply:
Yes, it probably should be an error.

Reply:
Actually I'm not sure we want to change this, an empty slice is technically a valid protobuf message with all fields set to the zero value. Maybe the right thing to do would be to set `v` to its zero value?